### PR TITLE
Initial version of simple FileReader/Writer

### PR DESCRIPTION
**Pull request by robert3005**
Reader currently assumes Chunked(Column(Chunked)) layouts. This is going to be expanded in follow ups

### CODE REVIEW

File: bench-vortex/src/reader.rs [hunk 1]
```diff
@@ -177,10 +179,25 @@ pub async fn take_vortex_object_store<O: ObjectStore>(
 pub async fn take_vortex_tokio(path: &Path, indices: &[u64]) -> VortexResult<Array> {
     let len = File::open(path)?.metadata()?.len();
     let indices_array = indices.to_vec().into_array();
-    let taken = read_vortex_footer_format(TokioAdapter(tokio::fs::File::open(path).await?), len)
+
+    let file = TokioAdapter(tokio::fs::File::open(path).await?);
+    let reader = FileReaderBuilder::new(file).with_length(len).build();
+
+    use futures::StreamExt;
```
Comment (robert3005) [inline bench-vortex/src/reader.rs]: should move it to the top

File: bench-vortex/src/reader.rs [hunk 2]
```diff
@@ -177,10 +179,25 @@ pub async fn take_vortex_object_store<O: ObjectStore>(
...
+        .collect::<Vec<_>>()
+        .await
+        .into_iter()
+        .collect::<VortexResult<Vec<_>>>()?;
+
+    let dtype = data[0].dtype().clone();
+
+    let data = ChunkedArray::try_new(data, dtype)?.into_array();
+
+    let taken = take(&data, &indices_array)?;
```
Comment (robert3005) [inline bench-vortex/src/reader.rs]: this take has to be on the stream of layouts using metadata tables

File: vortex-serde/src/file/file_writer.rs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+                .into(),
+                vec![
+                    chunk.byte_offsets.into_array(),
+                    byte_counts.into_array(),
+                    chunk.row_offsets.into_array(),
+                    row_counts.into_array(),
+                ],
+                len,
+                Validity::NonNullable,
+            )?;
```
Comment (robert3005) [inline vortex-serde/src/file/file_writer.rs:209]: This logic beyond extending chunks byte offsets is highly suspect. We can't reuse these tables in ChunkedArrayReader which is a bad thing. I wonder if we need to store multiple metadata tables since the offset tables are going to be 1 row longer than stat tables. OR the byte/row offsets live in layouts and these tables are pure metadata.

File: vortex-serde/src/file/file_reader.rs [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    pub fn build(self) -> FileReader<R> {
+        FileReader {
+            inner: self.inner,
+            len: self.len,
+        }
+    }
+}
+
+impl<R: VortexReadAt> FileReader<R> {
+    const FOOTER_READ_SIZE: usize = 8 * 1024 * 1024;
```
Comment (a10y) [inline vortex-serde/src/file/file_reader.rs]: worth documenting that this is sized to be optimal for s3?

File: vortex-serde/src/file/file_reader.rs [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+    pub async fn read_footer(&mut self) -> VortexResult<Footer> {
+        let read_size = Self::FOOTER_READ_SIZE.min(self.len().await as usize);
+        dbg!(read_size);
```
Comment (a10y) [inline vortex-serde/src/file/file_reader.rs]: reminder to remove

File: vortex-serde/src/file/file_writer.rs [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+use vortex_flatbuffers::WriteFlatBuffer;
+
+use crate::file::layouts::{ChunkedLayout, FlatLayout, Layout, StructLayout};
+use crate::flatbuffers::footer as fb;
+use crate::io::VortexWrite;
+use crate::messages::IPCSchema;
+use crate::writer::ChunkLayout;
+use crate::MessageWriter;
+
+pub const MAGIC_BYTES: [u8; 4] = *b"SP1R";
```
Comment (a10y) [inline vortex-serde/src/file/file_writer.rs]: should this be a play on vortex or spiral?

File: vortex-serde/src/file/file_reader.rs (hunk 5 above)
Comment (a10y) [reply vortex-serde/src/file/file_reader.rs]: also we should assert that the len() >= footer size

File: vortex-serde/src/io/read.rs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        buffer: BytesMut,
+    ) -> impl Future<Output = io::Result<BytesMut>> + Send;
+
+    // TODO(ngates): the read implementation should be able to hint at its latency/throughput
+    //  allowing the caller to make better decisions about how to coalesce reads.
+    fn performance_hint(&self) -> usize {
+        0
+    }
+
+    fn len(&self) -> impl Future<Output = u64>;
```
Comment (a10y) [inline vortex-serde/src/io/read.rs]: not part of this PR, but `size` is probably a more appropriate name for this, to match std::fs and object_store nomenclature

File: vortex-serde/src/file/footer.rs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+impl Footer {
+    pub fn leftovers_footer_offset(&self) -> usize {
+        (self.footer_offset - self.leftovers_offset) as usize
+    }
+
+    pub fn leftovers_schema_offset(&self) -> usize {
+        (self.schema_offset - self.leftovers_offset) as usize
+    }
+
+    pub async fn layout(&self) -> VortexResult<Layout> {
```
Comment (a10y) [inline vortex-serde/src/file/footer.rs]: this and `dtype()` below don't currently need to be async

File: vortex-serde/src/file/file_reader.rs [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+            layout,
+            dtype,
+            reader: Some(self.inner),
+            state: StreamingState::default(),
+            context: Default::default(),
+        })
+    }
+}
+
+pub struct FileReaderStream<R> {
```
Comment (a10y) [inline vortex-serde/src/file/file_reader.rs]: how do you imagine we implement pushdown on top of this? do we just attach some filter expressions to this that we evaluate inside `poll_next()` ?

File: vortex-serde/src/file/file_reader.rs [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+                    }
+                    Err(e) => return Poll::Ready(Some(Err(e))),
+                },
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (a10y) [inline vortex-serde/src/file/file_reader.rs]: I'd move this to a tests/ folder since it's really an integration test b/w FileReader and FileWriter.

File: vortex-serde/src/file/file_writer.rs (hunk 6 above)
Comment (robert3005) [reply vortex-serde/src/file/file_writer.rs]: Well, I had this originally long time ago and it stuck in my head. Should be VTX1 or something like that

File: vortex-serde/src/io/read.rs (hunk 7 above)
Comment (robert3005) [reply vortex-serde/src/io/read.rs]: We can call it size. Somehow I’m always thinking in terms of length but makes sense to differentiate logical from physical unit

File: vortex-serde/src/file/file_reader.rs (hunk 9 above)
Comment (robert3005) [reply vortex-serde/src/file/file_reader.rs]: I think we will do some level of pruning before ie layout will be able to do work before reading data. So depending on the operation being performed we will tell layouts whether we want to scan or take and then they can do some extra work before returning data. We’d have to integrate it into this state machine

**Comment by lwwmanning** [comment]
note that the new vortex_serde tests appear to take a LOOONG time in miri, probably worth adding `#[cfg_attr(miri, ignore)]` those (and maybe making "fast" variants for miri to run)

**Review by AdamGS — APPROVED** [review]

File: vortex-serde/src/file/file_reader.rs (hunk 10 above)
Comment (robert3005) [reply vortex-serde/src/file/file_reader.rs]: Agreed. However, for now it's hard to write these tests. Will start refactoring after this merges
